File: javascript/src/ir-flow.mjs
import { TevScriptError } from "./values.mjs";

function fail(path, message) {
  throw new TevScriptError("TEVS_IR_FLOW_INVALID", `${path}: ${message}`);
}

function binarySignature(operator, left, right, result) {
  if (["AND", "OR"].includes(operator)) return left === "Bool" && right === "Bool" && result === "Bool";
  if (["EQEQ", "NE"].includes(operator)) return result === "Bool" && (left === right || (left === "Rat" && right === "Rat"));
  if (["LT", "LE", "GT", "GE"].includes(operator)) return left === right && ["Int", "Rat"].includes(left) && result === "Bool";
  if (["PLUS", "MINUS"].includes(operator)) {
    return (left === "Int" && right === "Int" && result === "Int")
      || (left === "Rat" && right === "Rat" && result === "Rat")
      || (left === right && result === left && ["Vec2", "Vec3"].includes(result));
  }
  if (operator === "STAR") {
    return (left === "Int" && right === "Int" && result === "Int")
      || (left === "Rat" && right === "Rat" && result === "Rat")
      || (["Vec2", "Vec3"].includes(left) && right === "Rat" && result === left)
      || (["Vec2", "Vec3"].includes(right) && left === "Rat" && result === right);
  }
  if (operator === "SLASH") {
    return (left === "Rat" && right === "Rat" && result === "Rat")
      || (["Vec2", "Vec3"].includes(left) && right === "Rat" && result === left);
  }
  return false;
}

function maps(entity, handler) {
  return {
    states: new Map(entity.states.map((item) => [item.name, item.type])),
    parameters: new Map(handler.parameters.map((item) => [item.name, item.type])),
    locals: new Map(handler.locals.map((item) => [item.name, item.type])),
    capabilities: new Map(entity.capabilities.map((item) => [item.capability_id, {
      parameters: [...item.parameters], returnType: item.return_type, kind: item.kind,
    }])),
    events: new Map(entity.emitted_events.map((item) => [item.event_id, [...item.parameters]])),
  };
}

function sameStack(left, right) {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

export function validateEntityHandlerFlow(entity, handler, path) {
  const { states, parameters, locals, capabilities, events } = maps(entity, handler);
  const instructions = handler.instructions;
  const incoming = new Array(instructions.length).fill(null);
  incoming[0] = { stack: [], initialized: new Set() };
  let reachedReturn = false;

  function merge(target, stack, initialized, source) {
    if (!Number.isInteger(target) || target < 0 || target >= instructions.length) {
      fail(source, "control flow must target an instruction, not handler exit");
    }
    const next = { stack: [...stack], initialized: new Set(initialized) };
    const current = incoming[target];
    if (current === null) {
      incoming[target] = next;
      return;
    }
    if (!sameStack(current.stack, next.stack)) fail(source, `CFG merge stack mismatch at instruction ${target}`);
    incoming[target] = {
      stack: current.stack,
      initialized: new Set([...current.initialized].filter((name) => next.initialized.has(name))),
    };
  }

  for (let pc = 0; pc < instructions.length; pc += 1) {
    const state = incoming[pc];
    if (state === null) continue;
    const stack = [...state.stack];
    const initialized = new Set(state.initialized);
    const instruction = instructions[pc];
    const op = instruction.op;
    const currentPath = `${path}.instructions[${pc}]`;

    function pop(expected) {
      if (stack.length === 0) fail(currentPath, `stack underflow; expected ${expected}`);
      const actual = stack.pop();
      if (actual !== expected) fail(currentPath, `stack type mismatch; expected ${expected}, got ${actual}`);
    }

    if (op === "CONST") stack.push(instruction.type);
    else if (op === "LOAD_STATE") stack.push(states.get(instruction.name));
    else if (op === "STORE_STATE") pop(states.get(instruction.name));
    else if (op === "LOAD_LOCAL") {
      if (!initialized.has(instruction.name)) fail(currentPath, `local ${JSON.stringify(instruction.name)} is not definitely initialized`);
      stack.push(locals.get(instruction.name));
    } else if (op === "STORE_LOCAL") {
      pop(locals.get(instruction.name));
      initialized.add(instruction.name);
    } else if (op === "LOAD_PARAM") stack.push(parameters.get(instruction.name));
    else if (op === "CONVERT_INT_TO_RAT") { pop("Int"); stack.push("Rat"); }
    else if (op === "UNARY") {
      pop(instruction.operator === "NOT" ? "Bool" : instruction.type);
      stack.push(instruction.type);
    } else if (op === "BINARY") {
      if (!binarySignature(instruction.operator, instruction.left_type, instruction.right_type, instruction.result_type)) {
        fail(currentPath, "binary operator/type contract is not a V0.2 lowering");
      }
      pop(instruction.right_type);
      pop(instruction.left_type);
      stack.push(instruction.result_type);
    } else if (op === "CALL_PURE") {
      let expected;
      if (instruction.function_id === "vec2") expected = ["Rat", "Rat"];
      else if (instruction.function_id === "vec3") expected = ["Rat", "Rat", "Rat"];
      else if (["max", "min"].includes(instruction.function_id)) expected = [instruction.return_type, instruction.return_type];
      else fail(currentPath, `unknown pure function ${JSON.stringify(instruction.function_id)}`);
      [...expected].reverse().forEach(pop);
      if (instruction.return_type !== "Unit") stack.push(instruction.return_type);
    } else if (op === "CALL_CAPABILITY") {
      const signature = capabilities.get(instruction.capability_id);
      [...signature.parameters].reverse().forEach(pop);
      if (signature.returnType !== "Unit") stack.push(signature.returnType);
    } else if (op === "EMIT_EVENT") {
      [...events.get(instruction.event_id)].reverse().forEach(pop);
    } else if (op === "JUMP_IF_FALSE") {
      pop("Bool");
      merge(instruction.target, stack, initialized, currentPath);
      if (pc + 1 >= instructions.length) fail(currentPath, "conditional fallthrough exits the handler");
      merge(pc + 1, stack, initialized, currentPath);
      continue;
    } else if (op === "JUMP") {
      merge(instruction.target, stack, initialized, currentPath);
      continue;
    } else if (op === "RETURN") {
      if (stack.length !== 0) fail(currentPath, `RETURN requires empty stack, found ${JSON.stringify(stack)}`);
      reachedReturn = true;
      continue;
    } else fail(currentPath, `unknown opcode ${JSON.stringify(op)}`);

    if (pc + 1 >= instructions.length) fail(currentPath, "reachable control flow falls off the handler");
    merge(pc + 1, stack, initialized, currentPath);
  }
  if (!reachedReturn) fail(path, "no reachable RETURN");
}

File: tev_script/ir_flow.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .diagnostics import TevScriptError


@dataclass(frozen=True, slots=True)
class _FlowState:
    stack: tuple[str, ...]
    initialized_locals: frozenset[str]


def _fail(path: str, message: str) -> None:
    raise TevScriptError("TEVS_IR_FLOW_INVALID", f"{path}: {message}")


def _binary_signature(operator: str, left: str, right: str, result: str) -> bool:
    if operator in {"AND", "OR"}:
        return left == right == result == "Bool"
    if operator in {"EQEQ", "NE"}:
        return result == "Bool" and (
            left == right or (left == right == "Rat")
        )
    if operator in {"LT", "LE", "GT", "GE"}:
        return left == right and left in {"Int", "Rat"} and result == "Bool"
    if operator in {"PLUS", "MINUS"}:
        return (
            (left == right == result == "Int")
            or (left == right == result == "Rat")
            or (left == right == result and result in {"Vec2", "Vec3"})
        )
    if operator == "STAR":
        return (
            (left == right == result == "Int")
            or (left == right == result == "Rat")
            or (left in {"Vec2", "Vec3"} and right == "Rat" and result == left)
            or (right in {"Vec2", "Vec3"} and left == "Rat" and result == right)
        )
    if operator == "SLASH":
        return (
            (left == right == result == "Rat")
            or (left in {"Vec2", "Vec3"} and right == "Rat" and result == left)
        )
    return False


def _maps(entity: dict[str, Any], handler: dict[str, Any]) -> tuple[
    dict[str, str],
    dict[str, str],
    dict[str, str],
    dict[str, tuple[tuple[str, ...], str, str]],
    dict[str, tuple[str, ...]],
]:
    states = {str(item["name"]): str(item["type"]) for item in entity["states"]}
    parameters = {
        str(item["name"]): str(item["type"]) for item in handler["parameters"]
    }
    locals_ = {str(item["name"]): str(item["type"]) for item in handler["locals"]}
    capabilities = {
        str(item["capability_id"]): (
            tuple(str(value) for value in item["parameters"]),
            str(item["return_type"]),
            str(item["kind"]),
        )
        for item in entity["capabilities"]
    }
    events = {
        str(item["event_id"]): tuple(str(value) for value in item["parameters"])
        for item in entity["emitted_events"]
    }
    return states, parameters, locals_, capabilities, events


def validate_entity_handler_flow(
    entity: dict[str, Any],
    handler: dict[str, Any],
    path: str,
) -> None:
    states, parameters, locals_, capabilities, events = _maps(entity, handler)
    instructions = list(handler["instructions"])
    count = len(instructions)
    incoming: list[_FlowState | None] = [None] * count
    incoming[0] = _FlowState((), frozenset())
    reached_return = False

    def merge(target: int, stack: list[str], initialized: set[str], source: str) -> None:
        if target < 0 or target >= count:
            _fail(source, "control flow must target an instruction, not handler exit")
        state = _FlowState(tuple(stack), frozenset(initialized))
        current = incoming[target]
        if current is None:
            incoming[target] = state
            return
        if current.stack != state.stack:
            _fail(source, f"CFG merge stack mismatch at instruction {target}")
        incoming[target] = _FlowState(
            current.stack,
            current.initialized_locals & state.initialized_locals,
        )

    for pc, instruction in enumerate(instructions):
        state = incoming[pc]
        if state is None:
            continue
        stack = list(state.stack)
        initialized = set(state.initialized_locals)
        op = str(instruction["op"])
        current_path = f"{path}.instructions[{pc}]"

        def pop(expected: str) -> None:
            if not stack:
                _fail(current_path, f"stack underflow; expected {expected}")
            actual = stack.pop()
            if actual != expected:
                _fail(current_path, f"stack type mismatch; expected {expected}, got {actual}")

        if op == "CONST":
            stack.append(str(instruction["type"]))
        elif op == "LOAD_STATE":
            stack.append(states[str(instruction["name"])])
        elif op == "STORE_STATE":
            pop(states[str(instruction["name"])])
        elif op == "LOAD_LOCAL":
            name = str(instruction["name"])
            if name not in initialized:
                _fail(current_path, f"local {name!r} is not definitely initialized")
            stack.append(locals_[name])
        elif op == "STORE_LOCAL":
            name = str(instruction["name"])
            pop(locals_[name])
            initialized.add(name)
        elif op == "LOAD_PARAM":
            stack.append(parameters[str(instruction["name"])])
        elif op == "CONVERT_INT_TO_RAT":
            pop("Int")
            stack.append("Rat")
        elif op == "UNARY":
            result = str(instruction["type"])
            operator = str(instruction["operator"])
            expected = "Bool" if operator == "NOT" else result
            pop(expected)
            stack.append(result)
        elif op == "BINARY":
            left = str(instruction["left_type"])
            right = str(instruction["right_type"])
            result = str(instruction["result_type"])
            operator = str(instruction["operator"])
            if not _binary_signature(operator, left, right, result):
                _fail(current_path, "binary operator/type contract is not a V0.2 lowering")
            pop(right)
            pop(left)
            stack.append(result)
        elif op == "CALL_PURE":
            function_id = str(instruction["function_id"])
            return_type = str(instruction["return_type"])
            if function_id == "vec2":
                expected = ("Rat", "Rat")
            elif function_id == "vec3":
                expected = ("Rat", "Rat", "Rat")
            elif function_id in {"max", "min"}:
                expected = (return_type, return_type)
            else:
                _fail(current_path, f"unknown pure function {function_id!r}")
                raise AssertionError
            for type_name in reversed(expected):
                pop(type_name)
            if return_type != "Unit":
                stack.append(return_type)
        elif op == "CALL_CAPABILITY":
            capability_id = str(instruction["capability_id"])
            signature = capabilities[capability_id]
            for type_name in reversed(signature[0]):
                pop(type_name)
            if signature[1] != "Unit":
                stack.append(signature[1])
        elif op == "EMIT_EVENT":
            event_id = str(instruction["event_id"])
            for type_name in reversed(events[event_id]):
                pop(type_name)
        elif op == "JUMP_IF_FALSE":
            pop("Bool")
            merge(int(instruction["target"]), stack, initialized, current_path)
            if pc + 1 >= count:
                _fail(current_path, "conditional fallthrough exits the handler")
            merge(pc + 1, stack, initialized, current_path)
            continue
        elif op == "JUMP":
            merge(int(instruction["target"]), stack, initialized, current_path)
            continue
        elif op == "RETURN":
            if stack:
                _fail(current_path, f"RETURN requires empty stack, found {stack}")
            reached_return = True
            continue
        else:
            _fail(current_path, f"unknown opcode {op!r}")

        if pc + 1 >= count:
            _fail(current_path, "reachable control flow falls off the handler")
        merge(pc + 1, stack, initialized, current_path)

    if not reached_return:
        _fail(path, "no reachable RETURN")
